pkg/services/rendering/http_mode.go

```diff
@@ -83,42 +113,81 @@ func (rs *RenderingService) renderViaHttp(ctx context.Context, renderKey string,
 		}
 	}()
 
+	var downloadFileName string
+	_, params, err := mime.ParseMediaType(resp.Header.Get("Content-Disposition"))
+	if err == nil {
+		downloadFileName = params["filename"]
+	}
+
+	err = rs.readFileResponse(resp, filePath, reqContext)
+	if err != nil {
+		return nil, err
+	}
+
+	return &RenderCSVResult{FilePath: filePath, FileName: downloadFileName}, nil
+}
+
+func (rs *RenderingService) doRequest(url *url.URL, headers map[string][]string, ctx context.Context) (*http.Response, error) {
+	req, err := http.NewRequest("GET", url.String(), nil)
+	if err != nil {
+		return nil, err
+	}
+
+	req.Header.Set("User-Agent", fmt.Sprintf("Grafana/%s", setting.BuildVersion))
+	for k, v := range headers {
+		req.Header[k] = v
+	}
+
+	req = req.WithContext(ctx)
+
+	rs.log.Debug("calling remote rendering service", "url", url)
+
+	// make request to renderer server
+	resp, err := netClient.Do(req)
+	if err != nil {
+		rs.log.Error("Failed to send request to remote rendering service.", "error", err)
+		return nil, fmt.Errorf("failed to send request to remote rendering service: %w", err)
+	}
+
+	return resp, nil
+}
+
+func (rs *RenderingService) readFileResponse(resp *http.Response, filePath string, ctx context.Context) error {
 	// check for timeout first
-	if errors.Is(reqContext.Err(), context.DeadlineExceeded) {
+	if errors.Is(ctx.Err(), context.DeadlineExceeded) {
 		rs.log.Info("Rendering timed out")
-		return nil, ErrTimeout
+		return ErrTimeout
 	}
 
 	// if we didn't get a 200 response, something went wrong.
 	if resp.StatusCode != http.StatusOK {
 		rs.log.Error("Remote rendering request failed", "error", resp.Status)
-		return nil, fmt.Errorf("remote rendering request failed, status code: %d, status: %s", resp.StatusCode,
+		return fmt.Errorf("remote rendering request failed, status code: %d, status: %s", resp.StatusCode,
 			resp.Status)
 	}
 
 	out, err := os.Create(filePath)
 	if err != nil {
-		return nil, err
+		return err
 	}
+
 	defer func() {
 		if err := out.Close(); err != nil {
-			// We already close the file explicitly in the non-error path, so shouldn't be a problem
```

Comment:
Why remove this comment? I added it to make it clear why we close twice.

Reply:
I added it back but I also added a check to not log if the file is already closed as it happens all the time for successful rendering and it was a bit confusing.